api/datastore/datastore.go

```diff
@@ -28,34 +32,113 @@ func NewStore(storePath string, fileService portainer.FileService, connection po
-func (store *Store) Open() (newStore bool, err error) {
-	newStore = true
-	err = store.connection.Open()
+func (store *Store) Open() error {
+	err := store.connection.Open()
+	if err != nil {
+		return err
+	}
+
+	// Check if DB is encrypted
+	ok, err := store.connection.IsEncryptionRequired()
 	if err != nil {
-		return newStore, err
+		logrus.Warnf("Error calling IsEncryptionRequired: %s", err.Error())
+	}
+	logrus.Infof("Is migration required?: %t", ok)
+	if ok {
+		// unencrypted database
+		store.encryptDB()
 	}
 
+	// This creates the accessor structures, and should not mutate the data in any way
 	err = store.initServices()
 	if err != nil {
-		return newStore, err
+		return err
 	}
 
 	// if we have DBVersion in the database then ensure we flag this as NOT a new store
-	if _, err := store.VersionService.DBVersion(); err == nil {
-		newStore = false
+	version, err := store.VersionService.DBVersion()
+	logrus.Println(fmt.Sprintf("version: %d", version))
+	if err == nil {
+		store.isNew = false
+		logrus.WithField("version", version).Infof("Opened existing store")
+	} else {
+		if err.Error() == "encrypted string too short" {
+			logrus.WithError(err).Debugf("open db failed - %s", err.Error())
+			return err
+		}
+		if store.IsErrObjectNotFound(err) {
+			// TODO: Validate if needed
+			err := store.VersionService.StoreDBVersion(portainer.DBVersion)
+			if err != nil {
+				return err
+			}
+			// its new, lets see if there's an import.yml file, and if there is, import it
+			importFile := "/data/import.json"
+			if exists, _ := store.fileService.FileExists(importFile); exists {
+				if err := store.Import(importFile); err != nil {
+					logrus.WithError(err).Debugf("import %s failed", importFile)
+
+					// TODO: should really rollback on failure, but then we have nothing.
+				} else {
+					logrus.Printf("Successfully imported %s to new portainer database", importFile)
+				}
+			}
+		} else {
+			logrus.WithError(err).Debugf("open db failed")
+			return err
+		}
 	}
 
-	return newStore, nil
+	return nil
 }
 
 func (store *Store) Close() error {
 	return store.connection.Close()
 }
 
+func (store *Store) encryptDB() error {
+	// Since database is not encrypted so
+	// settings this flag to false will not
+	// allow connection to use encryption key
+	store.connection.SetIsDBEncryptedFlag(false)
+	err := store.initServices()
+	if err != nil {
+		logrus.Fatal("init services failed")
+	}
+
+	// export file path for backup
+	exportFilename := path.Join(store.databasePath() + "." + fmt.Sprintf("backup-%d.json", time.Now().Unix()))
+
+	logrus.Infof("exporting database backup to %s", exportFilename)
+	err = store.Export(exportFilename)
+	if err != nil {
+		logrus.WithError(err).Debugf("failed to export to %s", exportFilename)
+		return err
+	}
+
+	logrus.Infof("database backup exported")
+
+	// Set isDBEncryptedFlag to true to import JSON in the encrypted format
+	store.connection.SetIsDBEncryptedFlag(true)
+	err = store.Import(exportFilename)
```

Comment:
so this (should) over-write the original bolt file? I think that means we need to make a (non-export) backup to rollback to?

Reply:
Done